ap/inference/datacard.py

```diff
@@ -0,0 +1,550 @@
+                            direction="Down",
+                        )
+                        up_name = syst_pattern.format(
+                            category=cat_name,
+                            process=proc_name,
+                            parameter=param_obj.name,
+                            direction="Up",
+                        )
+                        out_file[down_name] = h_down
+                        out_file[up_name] = h_up
+
+                    # save the effect
+                    __effects[param_obj.name] = (
+                        safe_div(integral(h_down), integral(h_nom)),
+                        safe_div(integral(h_up), integral(h_nom)),
+                    )
+
+            # dedicated data handling
+            cat_obj = self.inference_model_inst.get_category(cat_name)
+            if cat_obj.data_datasets:
+                if "data" not in hists:
+                    raise Exception(
+                        f"the inference model '{self.inference_model_inst.name}' is configured to "
+                        f"use real data in category '{cat_name}' but no histogram named 'data' "
+                        "exists",
+                    )
+
+                # simply save the data histogram
+                h_data = hists["data"]["nominal"].copy()
+                data_name = data_pattern.format(category=cat_name)
+                out_file[data_name] = h_data
+                _rates["data"] = h_data.sum().value
+
+            elif cat_obj.data_from_processes:
+                # fake data from processes
+                h_data = [hists[proc_name]["nominal"] for proc_name in cat_obj.data_from_processes]
+                h_data = sum(h_data[1:], h_data[0].copy())
+                data_name = data_pattern.format(category=cat_name)
+                out_file[data_name] = h_data
+                _rates["data"] = h_data.sum().value
+
+        return (rates, effects, nom_pattern_comb, syst_pattern_comb)
+
+    @classmethod
+    def align_lines(
+        cls,
+        lines: Sequence[Any],
+        sep: str = "  ",
+    ) -> List[str]:
+        lines = [
+            (line.split() if isinstance(line, str) else list(map(str, law.util.flatten(line))))
+            for line in lines
+        ]
+
+        lengths = {len(line) for line in lines}
+        if len(lengths) > 1:
+            raise Exception(
+                f"line alignment cannot be performed with lines of varying lengths: {lengths}",
+            )
+
+        # convert to rows and get the maximum width per row
+        n_rows = list(lengths)[0]
+        rows = [
+            [line[j] for line in lines]
+            for j in range(n_rows)
+        ]
+        max_widths = [
+            max(len(s) for s in row)
+            for row in rows
+        ]
+
+        # stitch back
+        return [
+            sep.join((s + " " * (max_widths[j] - len(s))) for j, s in enumerate(line))
+            for i, line in enumerate(lines)
+        ]
```

Comment:
Unless I'm missing some datacard feature, having this configurable might be bit of an overkill :P

Reply:
I mean, true - it's definetly overkill to make the padding configurable :D The same is probably also true for the alignment itself. I still think it might be nice to use the python string formatting instead of multiplying the spaces, but this is obviously my personal opinion :D Btw, why is there an enumerate for the lines? we don't actually use the variable `i` here, do we?

Reply:
Fine for me. And no clue about the `enumerate(lines)` ^^.

Reply:
Update 👍